instrumentation-api-incubator/src/main/java/io/opentelemetry/instrumentation/api/incubator/builder/internal/DefaultHttpClientTelemetryBuilder.java

```diff
@@ -0,0 +1,185 @@
+
+import com.google.errorprone.annotations.CanIgnoreReturnValue;
+import io.opentelemetry.api.OpenTelemetry;
+import io.opentelemetry.context.propagation.TextMapSetter;
+import io.opentelemetry.instrumentation.api.incubator.semconv.http.HttpClientExperimentalMetrics;
+import io.opentelemetry.instrumentation.api.incubator.semconv.http.HttpClientPeerServiceAttributesExtractor;
+import io.opentelemetry.instrumentation.api.incubator.semconv.http.HttpExperimentalAttributesExtractor;
+import io.opentelemetry.instrumentation.api.incubator.semconv.net.PeerServiceResolver;
+import io.opentelemetry.instrumentation.api.instrumenter.AttributesExtractor;
+import io.opentelemetry.instrumentation.api.instrumenter.Instrumenter;
+import io.opentelemetry.instrumentation.api.instrumenter.InstrumenterBuilder;
+import io.opentelemetry.instrumentation.api.instrumenter.SpanKindExtractor;
+import io.opentelemetry.instrumentation.api.instrumenter.SpanNameExtractor;
+import io.opentelemetry.instrumentation.api.semconv.http.HttpClientAttributesExtractor;
+import io.opentelemetry.instrumentation.api.semconv.http.HttpClientAttributesExtractorBuilder;
+import io.opentelemetry.instrumentation.api.semconv.http.HttpClientAttributesGetter;
+import io.opentelemetry.instrumentation.api.semconv.http.HttpClientMetrics;
+import io.opentelemetry.instrumentation.api.semconv.http.HttpSpanNameExtractor;
+import io.opentelemetry.instrumentation.api.semconv.http.HttpSpanNameExtractorBuilder;
+import io.opentelemetry.instrumentation.api.semconv.http.HttpSpanStatusExtractor;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+import java.util.function.Consumer;
+import java.util.function.Function;
+
+/**
+ * This class is internal and is hence not for public use. Its APIs are unstable and can change at
+ * any time.
+ */
+public final class DefaultHttpClientTelemetryBuilder<REQUEST, RESPONSE> {
+
+  private final String instrumentationName;
+  private final OpenTelemetry openTelemetry;
+
+  private final List<AttributesExtractor<? super REQUEST, ? super RESPONSE>> additionalExtractors =
+      new ArrayList<>();
+  private final HttpClientAttributesExtractorBuilder<REQUEST, RESPONSE>
+      httpAttributesExtractorBuilder;
+  private final HttpClientAttributesGetter<REQUEST, RESPONSE> attributesGetter;
+  private final Optional<TextMapSetter<REQUEST>> headerSetter;
+  private final HttpSpanNameExtractorBuilder<REQUEST> httpSpanNameExtractorBuilder;
+  private Function<SpanNameExtractor<REQUEST>, ? extends SpanNameExtractor<? super REQUEST>>
+      spanNameExtractorTransformer = Function.identity();
+  private boolean emitExperimentalHttpClientMetrics = false;
+
+  public DefaultHttpClientTelemetryBuilder(
+      String instrumentationName,
+      OpenTelemetry openTelemetry,
+      HttpClientAttributesGetter<REQUEST, RESPONSE> attributesGetter,
+      Optional<TextMapSetter<REQUEST>> headerSetter) {
```

Comment:
how about moving this param out of the constructor and to a builder method that is only called when `TextMapSetter` is needed?

Reply:
this builder is reused among agent and library instrumentations.

By passing all required info in the ctor, we don't have to repeat the header setter in among agent and library instrumentations - or among several library instrumentations.

Furthermore, there are less types involved at the call site, making it easier to understand.

Reply:
I'm not sure if we're talking about the same thing(?)

typically builder constructors only take required parameters, and have normal methods on the builder to set the optional parameters

Reply:
an example - `JettyClientTelemetry`

usage in agent:

https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/51b3add44e181bc76c10ed9cbd54f901c0c7206c/instrumentation/jetty-httpclient/jetty-httpclient-9.2/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/jetty/httpclient/v9_2/JettyHttpClientSingletons.java#L17-L19

library:

https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/6dbad9abdac64ad228f49f8242e3a0937c1f4006/instrumentation/jetty-httpclient/jetty-httpclient-9.2/library/src/main/java/io/opentelemetry/instrumentation/jetty/httpclient/v9_2/JettyClientTelemetryBuilder.java#L133-L138

we should not have to pass `HttpHeaderSetter.INSTANCE` twice - in both snippets.

Maybe you mean something else?

Reply:
Maybe

```java
  builder =
    DefaultHttpClientTelemetryBuilder.create(INSTRUMENTATION_NAME,
                openTelemetry,
                JettyClientHttpAttributesGetter.INSTANCE).contextPropagator(HttpHeaderSetter.INSTANCE);
```

instead of 

https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/6dbad9abdac64ad228f49f8242e3a0937c1f4006/instrumentation/jetty-httpclient/jetty-httpclient-9.2/library/src/main/java/io/opentelemetry/instrumentation/jetty/httpclient/v9_2/JettyClientTelemetryBuilder.java#L34-L39